########################
## Basic parameters

## Sample information file with five columns: Id / Name / Group / Fq1 / Fq2 / Ctrl / PeakMode
src_sampleInfo	= "../sample.tsv"
cluster_yml	= "~/bin/CnR/Snakemake.bySample/Pool/cluster.yml"

## Name / PeakMode / Ctrl columns for group analysis mimicking sample.tsv file
## i.e. Name correspond to the Group column in sample.tsv
src_groupInfo	= "group.tsv"

## Genome folder depending on platform: CCHMC:HPC vs my desktop
import socket
hostname = socket.gethostname()	
if( hostname == "EA19-00359" ):
	# my desktop
	genomeFa	= "/Users/limc8h/Research/Common_Data/hg38/genome/genome.fa"
	chrom_size	= "/Users/limc8h/Research/Common_Data/hg38/chrom.sizes"
	peak_mask	= "/Users/limc8h/Research/Common_Data/hg38/ENCODE-blacklist.bed"
	baseDir		= "/Volumes"
else:
	# cluster system
	genomeFa	= "/data/limlab/Resource/GenomeData/hg/hg38/Genome/genome.fa"
	chrom_size	= "/data/limlab/Resource/GenomeData/hg/hg38/Genome/chrom.size"
	peak_mask	= "/data/limlab/Resource/GenomeData/hg/hg38/hg38-blacklist.v2.bed"
	baseDir		= "/data"

## Other essential pameters
#genome			= "hg38"
#adapter			= "AGATCGGAAGAGC"	# illumina universal adapter
#adapter			= "CTGTCTCTTATA"	# Nextera adapter for ATAC-seq or Cut&Tag
#trim_maxLen=100	## Maximum read length after trimming. ** NOT YET IMPLEMENTED **
#trim_minLen		= 20
#trim_minQual	= 20
chrRegexAll		= "^chr[0-9XY]+$|^dm-chr[0-9XYLR]+$"
chrRegexTarget	= "^chr[0-9XY]+$"
spikePrefix		= "dm-"


#########################
## Directories


bamDir_rep	= "../1.2.Align.filtered"
bamDir_pool	= "1.2.Align.pool"

## Starting BAM file directory for downstream analysis
bamDir		= bamDir_pool

## Downstream analysis directories
dedupDir	= "../Protrusion2/1.3.Align.dedup"
qcDir		= "2.1.QualityControl"

sampleDir	= "3.Sample"



################################




################################
## Loading sample Information
import pandas as pd
import sys
sampleAll = pd.read_csv(src_sampleInfo, sep="\t", comment="#", na_filter=False)
if not sampleAll.Id.is_unique:
	print( "Error: Id column in sample.tsv is not unique")
	sys.exit()
if not sampleAll.Name.is_unique:
	print( "Error: Name column in sample.tsv is not unique")
	sys.exit()

samples = pd.read_csv(src_groupInfo, sep="\t", comment="#", na_filter=False)
if not samples.Name.is_unique:
	print( "Error: Id column in sample.tsv is not unique")
	sys.exit()

#################################
## Cluster configuration file
#cluster = json.load(open("./cluster.json"))
import yaml
with open(os.path.expanduser(cluster_yml), 'r') as fh:
	cluster = yaml.load(fh)





#########################
## Rules start
rule all:
	input:
		## QC output
		expand(sampleDir + "/{sampleName}/QC/fragLen.dist.{ext}", sampleName=sampleList, ext=["txt","png"]),
		expand(sampleDir + "/{sampleName}/QC/enrich.acor.{ext}", sampleName=sampleList, ext=["txt","png"]),
		expand(sampleDir + "/{sampleName}/QC/base_freq.{ext}",  sampleName=sampleList, ext=["png","html"]),
		alignDir + "/alignStat.txt"
		#spikeinCntDir + "/spikein.txt"
		## bigWig files	
		expand(sampleDir + "/{sampleName}/igv.{fragment}.ctr.bw", sampleName=sampleList, fragment=["all","nfr","nuc"]),
		expand(sampleDir + "/{sampleName}/igv.1bp.{strand}.bw", sampleName=sampleListFactor, strand=["plus","minus"]),
		expand(sampleDir + "/{sampleName}/igv.1bp.corrected{kmer_pseudo}.{strand}.bw", sampleName=sampleListFactor, kmer_pseudo=kmer_pseudo, strand=["plus","minus"]),
		expand(sampleDir + "/{sampleName}/igv.allFrag.bw", sampleName=sampleList),
		## peak files
		expand(sampleDir + "/{sampleName}/HomerPeak.factor/peak.exBL.1rpm.bed", sampleName=sampleListFactor),
		expand(sampleDir + "/{sampleName}/HomerPeak.histone/peak.exBL.bed", sampleName=sampleListHistone),
		expand(sampleDir + "/{sampleName}/Motif/Homer.all/homerResults.html", sampleName=sampleListFactor),
		expand(sampleDir + "/{sampleName}/Motif/MEME.random5k/meme-chip.html", sampleName=sampleListFactor),
		expand(sampleDir + "/{sampleName}/HomerPeak.factor/heatmap.exBL.1rpm.png", sampleName=sampleListFactor),
		## Footprint
		expand(sampleDir + "/{sampleName}/Footprint.Homer.default/cnr.4.complete", sampleName=sampleListFactor),
		expand(sampleDir + "/{sampleName}/Footprint.Homer.corrected{kmer_pseudo}/cnr.4.complete", sampleName=sampleListFactor, kmer_pseudo=kmer_pseudo)

include: os.environ["LIMLAB_BASE"] + "/CnR/Snakemake.bySample/rules.post.smk"
include: os.environ["LIMLAB_BASE"] + "/CnR/Snakemake.bySample/Pool/rules.smk"
